File: 13-roman-to-integer.c
#include <stdio.h>
#include <string.h>

int romanToInt(char *s) {
  int size = strlen(s);

  static const int trans[] = {
      ['I'] = 1,   ['V'] = 5,   ['X'] = 10,   ['L'] = 50,
      ['C'] = 100, ['D'] = 500, ['M'] = 1000,
  };

  if (size == 1) {
    return trans[(unsigned char)s[0]];
  }

  int total = 0;
  for (int i = 0; i < size; i++) {
    int current = trans[(unsigned char)s[i]];
    int next = trans[(unsigned char)s[i + 1]];
    if (current < next) {
      total += next - current;
      i++;
    } else {
      total += current;
    }
  }

  return total;
}

typedef struct {
  char *input;
  int expected;
} TestCase;

TestCase test_cases[] = {
    {"I", 1},
    {"II", 2},
    {"III", 3},
    {"IV", 4},
    {"V", 5},
    {"VI", 6},
    {"IX", 9},
    {"X", 10},
    {"XL", 40},
    {"XLIX", 49},
    {"L", 50},
    {"LVIII", 58},
    {"XC", 90},
    {"XCIX", 99},
    {"C", 100},
    {"CD", 400},
    {"D", 500},
    {"CM", 900},
    {"M", 1000},
    {"MCMXCIV", 1994},
    {"MMMCMXCIX", 3999}, // Maximum valid value
    {"MDCCLXXVI", 1776},
    {"MCMLIV", 1954},
    {"MMXXV", 2025},
    {"XIV", 14},
    {"XIX", 19},
    {"XLII", 42},
    {"CDXLIV", 444},
    {"CMXCIX", 999},
    {"MMM", 3000},
    {"MMMDCCCLXXXVIII", 3888}, // A longer valid one
};

int num_tests = sizeof(test_cases) / sizeof(test_cases[0]);

int main() {
  int passed = 0;
  int failed = 0;

  for (int i = 0; i < num_tests; i++) {
    char *input = test_cases[i].input;
    int expected = test_cases[i].expected;

    printf("\nTesting: \"%s\"\n", input);

    int result = romanToInt(input);

    if (result == expected) {
      printf("   Input:    \"%s\"\n", input);
      printf("   Expected: %d\n", expected);
      printf("   Got:      %d\n", result);
      passed++;
    } else {
      printf("! FAILED\n");
      printf("   Input:    \"%s\"\n", input);
      printf("   Expected: %d\n", expected);
      printf("   Got:      %d\n", result);
      failed++;
      return 1;
    }
  }

  printf("\n= * 40 \n");
  printf("SUMMARY: %d passed, %d failed\n", passed, failed);
  if (failed == 0) {
    printf("🎉 All tests passed!\n");
  }

  return 0;
}

File: 13-roman-to-integer.go
package main

import (
	"fmt"
)

func romanToInt(s string) int {
	total := 0
	prev := 0
	for i := len(s) - 1; i >= 0; i-- {
		curr := 0
		switch s[i] {
		case 'I':
			curr = 1
		case 'V':
			curr = 5
		case 'X':
			curr = 10
		case 'L':
			curr = 50
		case 'C':
			curr = 100
		case 'D':
			curr = 500
		case 'M':
			curr = 1000
		}
		if curr < prev {
			total -= curr
		} else {
			total += curr
		}
		prev = curr
	}
	return total
}

func main() {
	testCases := []struct {
		input    string
		expected int
	}{
		{"I", 1},
		{"II", 2},
		{"III", 3},
		{"IV", 4},
		{"V", 5},
		{"VI", 6},
		{"IX", 9},
		{"X", 10},
		{"XL", 40},
		{"XLIX", 49},
		{"L", 50},
		{"LVIII", 58},
		{"XC", 90},
		{"XCIX", 99},
		{"C", 100},
		{"CD", 400},
		{"D", 500},
		{"CM", 900},
		{"M", 1000},
		{"MCMXCIV", 1994},
		{"MMMCMXCIX", 3999}, // Maximum valid Roman numeral
		{"MDCCLXXVI", 1776},
		{"MCMLIV", 1954},
		{"MMXXV", 2025},
		{"XIV", 14},
		{"XIX", 19},
		{"XLII", 42},
		{"CDXLIV", 444},
		{"CMXCIX", 999},
		{"MMM", 3000},
		{"MMMDCCCLXXXVIII", 3888}, // A longer valid one
	}

	passed := 0
	failed := 0

	for _, tc := range testCases {
		result := romanToInt(tc.input)

		if result == tc.expected {
			passed++
		} else {
			failed++
			fmt.Println("\nFAILED: \"%s\"\n  Expected: %d\n  Got:      %d", tc.input, tc.expected, result)
			panic("F")
		}
	}

	fmt.Printf("SUMMARY: %d passed, %d failed\n", passed, failed)
	if failed == 0 {
		fmt.Printf("🎉 All tests passed!\n")
	}
}
